### PR TITLE
add metrics

### LINKED ISSUE #46: add prometheus metrics
Add the following metrics:

- Total number of errors
- Total number of machines
- Total number of nodes
- Time how long it took to create/delete a instance on the cloud provider
- Timedifference between node.CreationTimestamp -
 machine.CreationTimestamp

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
It adds basic metrics

**Which issue(s) this PR fixes**:
Fixes #46, #61

### CODE REVIEW

File: pkg/controller/machine.go [hunk 1]
```diff
@@ -185,11 +212,36 @@ func (c *Controller) updateMachineError(machine *machinev1alpha1.Machine, reason
 	return machine, nil
 }
 
+func (c *Controller) getProviderInstance(prov cloud.Provider, machine *machinev1alpha1.Machine) (instance.Instance, error) {
+	start := time.Now()
+	defer c.metrics.ControllerOperation.With("operation", "get").Observe(time.Since(start).Seconds())
```
Comment (mrIncompetent) [inline pkg/controller/machine.go]: label should be renamed to `get-cloud-instance`

File: pkg/controller/machine.go [hunk 2]
```diff
@@ -185,11 +212,36 @@ func (c *Controller) updateMachineError(machine *machinev1alpha1.Machine, reason
...
+	return prov.Get(machine)
+}
+
+func (c *Controller) deleteProviderInstance(prov cloud.Provider, machine *machinev1alpha1.Machine) error {
+	start := time.Now()
+	defer c.metrics.ControllerOperation.With("operation", "delete").Observe(time.Since(start).Seconds())
```
Comment (mrIncompetent) [inline pkg/controller/machine.go]: label should be renamed to `delete-cloud-instance`

File: pkg/controller/machine.go [hunk 3]
```diff
@@ -185,11 +212,36 @@ func (c *Controller) updateMachineError(machine *machinev1alpha1.Machine, reason
...
+	return prov.Delete(machine)
+}
+
+func (c *Controller) createProviderInstance(prov cloud.Provider, machine *machinev1alpha1.Machine, userdata string) (instance.Instance, error) {
+	start := time.Now()
+	defer c.metrics.ControllerOperation.With("operation", "create").Observe(time.Since(start).Seconds())
```
Comment (mrIncompetent) [inline pkg/controller/machine.go]: label should be renamed to `create-cloud-instance`

File: pkg/controller/machine.go [hunk 4]
```diff
@@ -185,11 +212,36 @@ func (c *Controller) updateMachineError(machine *machinev1alpha1.Machine, reason
...
+	return prov.Create(machine, userdata)
+}
+
+func (c *Controller) validateMachine(prov cloud.Provider, machine *machinev1alpha1.Machine) error {
+	start := time.Now()
+	defer c.metrics.ControllerOperation.With("operation", "validate").Observe(time.Since(start).Seconds())
```
Comment (mrIncompetent) [inline pkg/controller/machine.go]: label should be renamed to `validate-machine`

File: cmd/controller/metrics.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+func NewMachineControllerMetrics() *MachineControllerMetrics {
+	namespace := "machine"
+	subsystem := "controller"
+
+	cm := &MachineControllerMetrics{
+		Machines: prometheus.NewGaugeFrom(prom.GaugeOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "machines",
+			Help:      "The number of currently managed machines",
```
Comment (alvaroaleman) [inline cmd/controller/metrics.go]: Maybe omit the `currently managed` as there is no way a machine can exist but not be managed?

File: cmd/controller/main.go [hunk 6]
```diff
@@ -134,3 +151,15 @@ func parseClusterDNSIPs(s string) ([]net.IP, error) {
 	}
 	return ips, nil
 }
+
+func serveMetrics() {
+	m := http.NewServeMux()
+	m.Handle(prometheusPath, promhttp.Handler())
```
[the remark this answers is not in the document]
Comment (metalmatze) [reply cmd/controller/main.go]: Yep. That is how I always do it for my own projects. Just have like an internal http server and add handlers for health, metrics etc to it.
I was also fine with having multiple servers.

**Comment by mrIncompetent** [comment]
@alvaroaleman @metalmatze 
PTAL

**Comment by alvaroaleman** [comment]
Two small points:

* Please remove the `healthListenAddress` var and flag
* Please add `prometheus.io/{port,scrape}" annotations and  `- -listen-address=0.0.0.0:8085` to the example deployment

Otherwise lgtm, awesome work!`

**Comment by metalmatze** [comment]
I will check that in a bit when I get to the airport.

**Comment by mrIncompetent** [comment]
@alvaroaleman done with 560cb5e and 15300cf
Very good catch!

**Comment by metalmatze** [comment]
Approved in the plane before taking off in Brussels. 😂

**Pull request merged**